flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/SubstringIndexFunction.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.functions.scalar;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.data.StringData;
+import org.apache.flink.table.data.binary.BinaryStringData;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.table.functions.SpecializedFunction;
+import org.apache.flink.util.FlinkRuntimeException;
+
+import org.apache.commons.lang3.StringUtils;
+
+import javax.annotation.Nullable;
+
+import java.util.Arrays;
+
+/** Implementation of {@link BuiltInFunctionDefinitions#SUBSTRING_INDEX}. */
+@Internal
+public class SubstringIndexFunction extends BuiltInScalarFunction {
+    public SubstringIndexFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.SUBSTRING_INDEX, context);
+    }
+
+    public @Nullable StringData eval(
+            @Nullable StringData expr, @Nullable StringData delim, @Nullable Integer count) {
+        if (expr == null || delim == null || count == null) {
+            return null;
+        }
+
+        if (count == 0) {
+            return BinaryStringData.EMPTY_UTF8;
+        }
+        String str = expr.toString();
```

Comment:
Can use the BinaryStringData to judge and find the substring instead of materialized it in advance?

Reply:
Currently we don't have `find` and `rfind` for BinaryStringData. I think it would be great to introduce them into BinaryStringDataUtil  because they are commonly used for string matching.